capnp/src/private/arena.rs

```diff
@@ -288,54 +307,37 @@ where
     A: Allocator,
 {
     /// Allocates a new segment with capacity for at least `minimum_size` words.
-    fn allocate_segment(&mut self, minimum_size: WordCount32) -> Result<()> {
-        let seg = match &mut self.allocator {
-            Some(a) => a.allocate_segment(minimum_size),
-            None => unreachable!(),
-        };
+    fn allocate_segment(&mut self, minimum_size: WordCount32) -> SegmentId {
+        let seg = self.allocator.allocate_segment(minimum_size);
+        let segment_id = self.segments.len() as u32;
         self.segments.push(BuilderSegment {
             ptr: seg.0,
             capacity: seg.1,
-            allocated: 0,
+            allocated: minimum_size,
```

Comment:
This looks wrong to me. `allocated` is supposed to indicate how many words of the segment are already in use, which should be zero when the segment is first allocated. It seems to me that what you have here would result in a lot of unused memory.

I pushed some new doc comments for `BuilderSegment` that might make things clearer: https://github.com/capnproto/capnproto-rust/commit/888844337f000aeebd7b8fbb7d1bfcc6e728acd5

I suppose one things that's potentially confusing about all this is that the word "allocate" is being used for different things:
  1. the `message::Allocator`, whose job is to provide memory from the system
  2. `BuilderArena::allocate()`, which takes takes memory already-"allocated" memory from (1) and reserves it for usage by `private::layout`.

Reply:
Oh, I see. This has to do with commit https://github.com/capnproto/capnproto-rust/pull/362/commits/b6f69313d375fcdeb16c44a015b7f14deaff4881. It happens to work because there are only a two usages of the `allocate_segment()` method, and they follow a particular pattern.

I'd prefer not to make this change. If we do make it, we should at least add some comments about the new behavior, and maybe also change some naming.

Reply:
Maybe we can introduce a dedicated method for this then? I think having a non fallible method helps with readability and optimizations